torch/csrc/autograd/engine.cpp

```diff
@@ -1117,14 +1117,37 @@ void Engine::add_thread_pool_task(const std::weak_ptr<GraphTask>& graph_task) {
 }
 
 void GraphTask::init_to_execute(Node& graph_root, const edge_list& outputs, bool accumulate_grad) {
-  exec_info_[&graph_root].needed_ = true;
+  // Populates exec_info so nodes that should be executed have `exec_info[node].needed_ = true`
+  // Only nodes that have a path to any edge in `outputs` should be executed.
+  // The code below populates exec_info using recursion, but the actual code does this
+  // iteratively. Refer to the numbering to see how the actual code corresponds.
+  //
+  // is_needed = {fn: True for fn in outputs}             # (0)
+  // seen = {}
+  // def compute_is_needed(fn):
+  //   for next_edge in fn.next_edges:
+  //     child_fn = next_edge.fn
+  //     if child_fn in seen and is_needed[child_fn]:     # (1)
+  //       is_needed[fn] = true
+  //     else:
+  //       seen.add(child_fn)
+  //       if compute_is_needed(child_fn):
+  //         is_needed[fn] = true                         # (2)
+  //                                                      # (3)
```

Comment:
This Python function needs `return is_needed[fn]` at the end, because of `if compute_is_needed(child_fn)`.

Reply:
Addressed!